V2 - React Query + Unit Testing

Salman-Jara opened the pull request:
In previous push there were some problems in API calls. Fixed these.

React query added in:
- AQI Card
- News Card

Vitest Unit tests added for the card components

Salman-Jara committed:
Changed the API Calls in AQI and News component from simple fetch to react Query. Also added unit tests using vitest

Salman-Jara committed:
react query update

Salman-Jara committed:
react-query

Salman-Jara committed:
updated package.json file

Salman-Jara committed:
Version2 - React Query + Unit Tests

shahbazshueb commented on components/image-slideshow.tsx:
```diff
@@ -4,65 +4,40 @@ import { useEffect, useState } from "react";
-    setLoading(true)
-    
-    const url = "/api/fetchNewsData"
+  const [title, setTitle] = useState<string>("Pictures from space show mighty smog choking Lahore")
```
It is not necessary to mention type <string> here as default value's type is used which is already string here. Same thing for below.

shahbazshueb commented on components/aqi-card.tsx:
```diff
@@ -41,81 +42,59 @@ type OutsideCondition = {
-          setCondition({ description: "Hazardous air quality - try to stay indoors", icon: faMaskVentilator });
-        }
-        onSelect(aqi);
+  useEffect(() => {
```
Unnecessary call of useEffect. Remove this useEffect. 
Hint: Use Select with useQuery to resolve this.

shahbazshueb commented on components/aqi-card.tsx:
```diff
@@ -41,81 +42,59 @@ type OutsideCondition = {
-    } finally {
-      setLoading(false);
+
+      onSelect(data.aqi);
```
This callback which eventually sets state in the parent component is unnecessary. Simply call the same useAQIData hook where you need this data

shahbazshueb commented on tests/components/aqi-card.test.tsx:
```diff
@@ -0,0 +1,43 @@
+describe('AQICard', () => {
+  const mockOnSelect = vi.fn()
+
+  it('renders without crashing', () => {
```
The test cases here are very limited. There should be specific test cases for 
```
Unhealthy air quality - health side effects
```

```
Temperature: 19°C

Humidity: 37%
```

shahbazshueb replied:
I would suggest mocking the responses of API to test this out.

shahbazshueb replied:
Check out https://www.npmjs.com/package/nock

Salman-Jara committed:
Comments Resolution

- Removed the Unnecessary call of useEffect in AQICard
- Using customHook instead of Passing OnSelect to AqiCard component
